### ISSUE TITLE
What should nose2's support policy be now that python2 is EOL?

### ISSUE THREAD

**Issue by sirosen** [nose-devs/nose2#481]
Our current line in the docs reads

> nose2 only supports Python versions currently supported by the Python team.

but given that the primary users of nose2 are probably legacy use-cases, dropping python2 seems like a bad move which leaves those users without a viable non-unittest framework.

Developers will still be dealing with python2 for years, and though nose2 will never be as popular as pytest, supporting py2 may be a unique draw for nose2.

If we remove deprecated python3 versions and keep 2.7 as a special case, we'll reduce our testing matrix size and possibly make things easier on ourselves. In cases where unittest differs between 2.7, 3.5, and 3.6+, for example, we get to only support 2 variants instead of 3.
Even though supporting 3.5 and 3.4 may have a marginal cost today, I'm concerned about the ongoing cost we may incur down the line, and an inability to drop py3 versions as they reach EOL.

So right now, I'm inclined to set the policy as "python 2.7 and python 3 versions supported by the Python team".
I'm also inclined to remove jython from the tox matrix (since we don't run that on Travis) and perhaps to say something about pypy in the policy statement.

**Comment by stefanholek** [comment]
+1 on staying compatible with Python 2.7. There still is a lot of Python 2 code out there, and people going through porting efforts need a working test runner.

Jython no longer works under tox because virtualenv has stopped supporting it out of the box.

**Comment by hugovk** [comment]
Here's the pip installs for nose2 from PyPI for July 2021:

| category | percent | downloads |
| -------- | ------: | --------: |
| 3.7      |  61.57% |   296,297 |
| 3.8      |  16.05% |    77,224 |
| 2.7      |   8.41% |    40,451 |
| 3.6      |   6.68% |    32,161 |
| 3.9      |   5.00% |    24,058 |
| null     |   1.69% |     8,140 |
| 3.5      |   0.39% |     1,881 |
| 3.4      |   0.20% |       977 |
| 3.10     |   0.00% |        15 |
| 3.11     |   0.00% |         1 |
| Total    |         |   481,205 |

Source: `pip install -U pypistats && pypistats python_minor nose2 --last-month`

And all available data:

![image](https://user-images.githubusercontent.com/1324225/130230283-618fb747-dcc3-45ee-999f-a01d7458c69c.png)

A couple of things to note:

* Python 2.7 use is declining steadily, fairly typical (compare others: https://hugovk.github.io/pypi-tools/)
* Right now 3.6 is the oldest supported by the CPython core team, and there's _very little_ use of 3.5 and older, less than 1%

**Comment by sirosen** [comment]
@hugovk, thanks a ton for pulling together this data!

I've been operating on the assumption that most of nose2's users would be legacy codebases under py2.7 . This helps me reevaluate.

Do you know if the bigtable data can show us pypy or pypy3 info? Obviously, if pip under cpython downloads a package and then it runs under pypy, we'd never know. But I'm curious, since you've clearly got some experience in this space, if you've been able to put together anything usable for that?

**Comment by hugovk** [comment]
Yes, it's possible using https://github.com/ofek/pypinfo :

```console
$ pypinfo --start-date 2021-07-01 --end-date 2021-07-31 --percent --limit 100 nose2 pyversion impl
Served from cache: False
Data processed: 977.31 MiB
Data billed: 978.00 MiB
Estimated cost: $0.01

| python_version | implementation | percent | download_count |
| -------------- | -------------- | ------- | -------------- |
| 3.7            | CPython        |  62.72% |        296,210 |
| 3.8            | CPython        |  16.35% |         77,200 |
| 2.7            | CPython        |   8.53% |         40,306 |
| 3.6            | CPython        |   6.66% |         31,457 |
| 3.9            | CPython        |   5.09% |         24,030 |
| 3.5            | CPython        |   0.40% |          1,870 |
| 3.4            | CPython        |   0.21% |            973 |
| 3.6            | PyPy           |   0.03% |            136 |
| 2.7            | PyPy           |   0.02% |             91 |
| 3.10           | CPython        |   0.00% |             15 |
| 3.5            | PyPy           |   0.00% |              1 |
| 3.11           | CPython        |   0.00% |              1 |
| Total          |                |         |        472,290 |
```

**Comment by sirosen** [comment]
I've closed this, with the decision that I'm going to leave py2 support in for now.
I checked just now using `pypistats python_minor nose2 -t`, and it looks like 10% of downloads are still on py2.

I'll drop py2 support when it becomes a more serious threat to maintainability or if the usage numbers decline further.

I'd like to move to py3, but I do so little work on nose2 these days that it doesn't make sense to break people without a steady stream of improvements to make it worthwhile.
